Firestore/core/test/unit/local/remote_document_cache_test.cc

```diff
@@ -230,34 +232,85 @@ TEST_P(RemoteDocumentCacheTest, DocumentsMatchingUsesReadTimeNotUpdateTime) {
+  // data to ensure that the documents in the cache cannot be modified.
+  persistence_->Run("test_does_not_apply_document_modifications_to_cache", [&] {
+    MutableDocument document = SetTestDocument("coll/doc", Map("value", "old"));
+    document = cache_->Get(Key("coll/doc"));
+    VerifyValue(document, Map("value", "old"));
+    document.ConvertToFoundDocument(Version(kVersion),
+                                    ObjectValue{Map("value", "new")});
+
+    document = cache_->Get(Key("coll/doc"));
+    VerifyValue(document, Map("value", "old"));
+    document.ConvertToFoundDocument(Version(kVersion),
+                                    ObjectValue{Map("value", "new")});
+
+    MutableDocumentMap documents =
+        cache_->GetAll(DocumentKeySet{Key("coll/doc")});
+    document = documents.find(Key("coll/doc"))->second;
+    VerifyValue(document, Map("value", "old"));
+    document.ConvertToFoundDocument(Version(kVersion),
+                                    ObjectValue{Map("value", "new")});
+
+    documents = cache_->GetMatching(Query("coll"), SnapshotVersion::None());
+    document = documents.find(Key("coll/doc"))->second;
+    VerifyValue(document, Map("value", "old"));
+    document.ConvertToFoundDocument(Version(kVersion),
+                                    ObjectValue{Map("value", "new")});
+
+    document = cache_->Get(Key("coll/doc"));
+    VerifyValue(document, Map("value", "old"));
+  });
+}
 // MARK: - Helpers
 
-Document RemoteDocumentCacheTest::SetTestDocument(const absl::string_view path,
-                                                  int update_time,
-                                                  int read_time) {
-  Document doc = Doc(path, update_time, kDocData);
+MutableDocument RemoteDocumentCacheTest::SetTestDocument(
+    const absl::string_view path,
+    google_firestore_v1_Value data,
+    int update_time,
+    int read_time) {
+  MutableDocument doc = Doc(path, update_time, data);
   cache_->Add(doc, Version(read_time));
   return doc;
 }
 
-Document RemoteDocumentCacheTest::SetTestDocument(
+MutableDocument RemoteDocumentCacheTest::SetTestDocument(
+    const absl::string_view path, int update_time, int read_time) {
+  return SetTestDocument(path, DeepClone(kDocData), update_time, read_time);
+}
+
+MutableDocument RemoteDocumentCacheTest::SetTestDocument(
+    const absl::string_view path, google_firestore_v1_Value data) {
+  return SetTestDocument(path, data, kVersion, kVersion);
+}
+
+MutableDocument RemoteDocumentCacheTest::SetTestDocument(
     const absl::string_view path) {
-  return SetTestDocument(path, kVersion, kVersion);
+  return SetTestDocument(path, DeepClone(kDocData), kVersion, kVersion);
+}
+
+void RemoteDocumentCacheTest::VerifyValue(MutableDocument actual_doc,
+                                          google_firestore_v1_Value data) {
+  MutableDocument expected_doc =
+      Doc(actual_doc.key().ToString(), kVersion, data);
+  ASSERT_EQ(expected_doc, actual_doc);
```

Comment:
Can you use `EXPECT` rather than `ASSERT`? Also, if the failure happens in `VerifyValue`, it won't be obvious which invocation triggered it. You can either use [`SCOPED_TRACE`](https://github.com/google/googletest/blob/master/docs/advanced.md#adding-traces-to-assertions) before every invocation or make `VerifyValue` a matcher.

Reply:
I added a customer matcher.